### DISCUSSION TITLE
Firmware update question

### DISCUSSION THREAD

**Discussion by WesWesthaver** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#678] [Q&A]
Today I brought my SaluSpa Paris out of storage and got it ready for this new season. After setting everything up I thought I'd check to see if there were any firmware upgrades available but this is all I see when performing a "Check firmware update":

[ScreenCap](http://www.westhaver.com/Screenshot_2024-03-12_at_16-45-04_Lay-Z-Spa_Module_Check_firmware_update!.png)

Any idea why the Master and Beta fields are asking me to reload the page? (Btw, reloading the page yields the same result.)

**Comment by visualapproach** [comment]
That feature is withdrawn unfortunately. You need to download from GitHub and flash OTA or via cable.

**Reply by WesWesthaver** [reply]
Is this procedure documented anywhere? OTA update would be preferable.

**Reply by visualapproach** [reply]
Yes, follow the manual as you (probably) did at first install. OTA is possible by toggling comments on a few lines in platformio.ini (also in the manual)

**Reply by WesWesthaver** [reply]
As I recall, you use Visual Studio to handle the code development and firmware uploading etc. I purchased the controller module pre-built with firmware already installed so I don't have or use Visual Studio. I am a Ubuntu Linux user and am not sure if Visual Studio is possible on my system.

**Reply by visualapproach** [reply]
Ok, maybe there is a ESP flashing program for Ubuntu? In that case I (or someone else) could upload the FW file.

**Reply by WesWesthaver** [reply]
I found this: [info](https://blog.anavi.technology/getting-started-with-esptool-for-flashing-firmware-on-esp8266-and-esp32/) 

Looks like "[esptool](https://github.com/espressif/esptool)" should do the trick. I've installed it on my Ubuntu system. Can you provide the bin file needed to bring my spa controller up to date?

**Reply by peterdedecker83** [reply]
Visual Studio and its plugins work nice on Ubuntu Linux as well. I can update over cable or OTA and see the progress bar and success message, however, the device sticks to run a version of a few years old...

**Reply by WesWesthaver** [reply]
Thanks for the info. I swore off of MS products about 7 years ago after having spent my entire career in the MS camp. But it looks like VS for Linux may be ok (since it's open source.) I read that it's a light-weight version (feature wise) compared to the Windows version. I may install it and give it a spin. 

Not sure what you mean here:
> however, the device sticks to run a version of a few years old...

Can you clarify?

**Reply by peterdedecker83** [reply]
Well, I have an ESP running this software since a few years, which I wanted to update. So in VS Code, I pulled the latest from Github and clicked the arrow icon on the tray, which compiles the software and uploads it to the ESP, giving the necessary success messages. However, when rebooting the ESP and visiting the web interface, I still see the old version of a few years ago.

But in the meantime, I figured out that manually clicking the necessary steps separately (first build, then upload) actually results in an updated ESP running the latest version.

The readme.md file also mentions version 4.0.0-beta instead of the newest version.

**Reply by visualapproach** [reply]
You also need to build and upload the file system where all the html files lives. Or you can upload each file from the web ui.
